### PR TITLE
Start working on our own version of `cfg(target_has_atomic)`

### PR DESCRIPTION
**Pull request by myrrlyn**
I've been unhappy with our vague guessing of what atomics are available, and the #3 issue shows a target completely outside the scope of our temporary patch.

This branch uses a build script to inspect the `TARGET` environment variable, which Cargo helpfully sets to the value of the `--target ARCH` flag. This will be one of the targets enumerated in `rustup target list`.

I am running this commit through the `bitvec` cross-compile suite. Given that this suite is *extremely slow*, I expect that I will update with new targets as I discover their failures.

In the meantime, I would like review on whether we think this is an appropriate solution to the problem, and how we can make the script easily accessible to users to update with their architectures as needed.

### CODE REVIEW

**Comment by myrrlyn** [comment]
My `just cross` task has completed successfully with these commits. I don't feel like searching out the rest of the targets.

**Comment by myrrlyn** [comment]
This is driving me insane. I think I may actually bite the bullet and implement `available()` in the compiler just so I can throw all of this work away.

**Review by mystor — CHANGES_REQUESTED** [review]
If we're going to need the build script anyway, as the information isn't available from the `target_arch` config variable, let's just do it all in the build script. No reason to have the information split across two sources.

Sorry about asking for you to switch to using macros. I didn't realize that it wasn't possible to implement without a build script >_>

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
+//! Target detection
+//!
+//! Prefer putting target information in the `src/has_atomic.rs:has_atomic!`
+//! macro instead. Some targets, such as `riscv32imc`, are not uniquely
+//! selectable through the available compiler `cfg` discovery APIs, and so must
+//! be supported here. Where your target *is* precisely selectable through the
+//! existing `cfg` infrastructure, it belongs in the macro.
```
Comment (mystor) [inline build.rs:7]: The macro doesn't exist anymore, so let's drop it from the comment.

File: build.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    let arch = target.split('-').next().ok_or("Invalid target triple")?;
+    // Add new architecture sections here with their atomic availability.
+    #[allow(clippy::match_single_binding, clippy::single_match)]
+    match arch {
+        // "riscv32imc-unknown-none-elf" and "riscv32imac-unknown-none-elf" are
+        // both `target_arch = "riscv32", and have no `cfg`-discoverable
+        // distinction. As such, the atomic RISC-V target must be discovered
+        // here, rather than in the macro.
```
Comment (mystor) [inline build.rs:84]: The macro doesn't exist anymore :-)

File: src/lib.rs [hunk 3]
```diff
@@ -524,15 +534,16 @@ macro_rules! radium {
...
-            radium!(atom $base);
-            radium!(atom_bit $base);
-            radium!(atom_int $base);
-        }
+        #[cfg($flag)]
+		impl Radium<$base> for $atom {
+			radium!(atom $base);
+			radium!(atom_bit $base);
+			radium!(atom_int $base);
+		}
```
Comment (mystor) [inline src/lib.rs:546]: Please fix the indentation of these lines.

File: src/lib.rs [hunk 4]
```diff
@@ -772,4 +773,14 @@ mod tests {
...
+
+    #[test]
+    fn always_cell() {
+        static_assertions::assert_impl_all!(Cell<bool>: Radium<bool>);
+        static_assertions::assert_impl_all!(Cell<u8>: Radium<u8>);
+        static_assertions::assert_impl_all!(Cell<u16>: Radium<u16>);
+        static_assertions::assert_impl_all!(Cell<u32>: Radium<u32>);
+        static_assertions::assert_impl_all!(Cell<u64>: Radium<u64>);
+        static_assertions::assert_impl_all!(Cell<usize>: Radium<usize>);
+    }
```
Comment (mystor) [inline src/lib.rs:785]: We could also do this with the types from `types::RadiumT` :-)

**Review by mystor — APPROVED** [review]

File: build.rs (hunk 1 above)
Comment (myrrlyn) [reply build.rs:7]: lol I love to "update documentation"

File: src/lib.rs (hunk 3 above)
Comment (myrrlyn) [reply src/lib.rs:546]: This is what I get for not leaving `cargo watch -x fmt` running while I work, thanks VSC

**Pull request merged**